images/virtualization-artifact/pkg/controller/vm/internal/pod.go

```diff
@@ -0,0 +1,87 @@
+/*
+Copyright 2024 Flant JSC
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+     http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package internal
+
+import (
+	"context"
+
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	"github.com/deckhouse/virtualization-controller/pkg/controller/powerstate"
+	"github.com/deckhouse/virtualization-controller/pkg/controller/service"
+	"github.com/deckhouse/virtualization-controller/pkg/controller/vm/internal/state"
+	virtv2 "github.com/deckhouse/virtualization/api/core/v1alpha2"
+)
+
+const namePodHandler = "PodHandler "
+
+func NewPodHandler(client client.Client) *PodHandler {
+	return &PodHandler{
+		client:     client,
+		protection: service.NewProtectionService(client, virtv2.FinalizerPodProtection),
+	}
+}
+
+type PodHandler struct {
+	client     client.Client
+	protection *service.ProtectionService
+}
+
+func (h *PodHandler) Handle(ctx context.Context, s state.VirtualMachineState) (reconcile.Result, error) {
+	if s.VirtualMachine().IsEmpty() {
+		return reconcile.Result{}, nil
+	}
+	current := s.VirtualMachine().Current()
+	pods, err := s.Pods(ctx)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	if isDeletion(current) {
+		objs := make([]client.Object, len(pods.Items))
+		for i, p := range pods.Items {
+			objs[i] = p.DeepCopy()
+		}
+		return reconcile.Result{}, h.protection.RemoveProtection(ctx, objs...)
```

Comment:
I feel that we can pass pods.Items as-is without DeepCopy.

Reply:
We should add a pointer to objs, but we cannot add &p.
```
pkg/controller/vm/internal/pod.go:58:15: exporting a pointer for the loop variable p (exportloopref)
                        objs[i] = &p

```
We need to copy the object. And we have two ways
1. 
```go
pod = p 
objs[i] = &pod
```
2.
```go
objs[i] = p.DeepCopy()
```
but the first way does not perform a  deep copy.

Reply:
Oh, that's about array casting. May be create h.protection.RemoveProtectionFromPods to range []v1.Pod list and apply `h.protection.RemoveProtection` there without DeepCopy?

Reply:
fixed